go-controller/pkg/ovn/subnetallocator/allocator.go

```diff
@@ -55,35 +55,57 @@ func (sna *SubnetAllocator) MarkAllocatedNetwork(subnet *net.IPNet) error {
 	return fmt.Errorf("network %s does not belong to any known range", subnet.String())
 }
 
-func maybeAllocateOneNetwork(ranges []*subnetAllocatorRange, networks []*net.IPNet) ([]*net.IPNet, error) {
-	if len(ranges) == 0 {
-		return networks, nil
+// AllocateNetworks tries to allocate networks in all the ranges available
+func (sna *SubnetAllocator) AllocateNetworks() ([]*net.IPNet, error) {
+	var networks []*net.IPNet
+	var err error
+	ipv4network, err := sna.AllocateIPv4Network()
+	if err != nil {
+		return nil, err
+	}
+	if ipv4network != nil {
+		networks = append(networks, ipv4network)
+	}
+	ipv6network, err := sna.AllocateIPv6Network()
+	if err != nil {
+		return nil, err
+	}
+	if ipv6network != nil {
+		networks = append(networks, ipv6network)
+	}
+	return networks, nil
+}
+
+// AllocateIPv4Network tries to allocate an IPv4 network if there are ranges available
+func (sna *SubnetAllocator) AllocateIPv4Network() (*net.IPNet, error) {
+	sna.Lock()
+	defer sna.Unlock()
+	if len(sna.v4ranges) == 0 {
+		return nil, nil
```

Comment:
this is exactly the same behaviour it was implemented:

if the allocator doesn't have any "subnet range" configured where it can allocate subnets it returns nil

ErrSubnetAllocatorFull is returned if it runs out of subnets from the configured "subnet ranges"

Reply:
> this is exactly the same behaviour it was implemented:

Yes, I saw that, but it's still contradicting itself. I mean if `len(sna.v4ranges) == 0` is true we wouldn't loop and we'd return `ErrSubnetAllocatorFull`. So why are we returning `nil` instead? Moreover, the caller then proceeds to avoiding the nil value, which tells me we could just remove the check
```
if len(sna.v4ranges) == 0 {
		return nil, nil
}
```
 altogether.

Reply:
let´s assume you have configured a 10.0.0.0/16 and that you assign /24, you return ErrSubnetAllocatorFull when you exhausted all range because all the 255 10.0.x.0/24 subnets were allocated.
Let´s assume that you don´t have configured any IPv4 range in the allocator and you request a subnet, it returns nil.

Those were the semantic implemented, we shouldn't do changes on the semantics on refactors.
This commits refactors the allocator to, in addition to AllocateNetworks(), add two new methods AllocateNetworksIPv4() and AllocateNetworksIPv6()
The public methods maintain the initial semantic so we are sure we don't break any consumers of the allocator.

Reply:
I think both points are valid here. The original code is very awkward. Even the name "maybeAllocateOneNetwork" is ambiguous and in the code it reassigns the value of "networks" in the AllocateNetworks function, depending on the "maybeAllocateOneNetwork" function append to existing networks. Seems very out of place to me. Really the caller should call "AllocateOneNetwork" only if the sna.v4ranges !=nil, and not leave ambiguity into the callee. Either way we need to fix this I believe. If @aojea wants to retain the current state to be sure he doesnt introduce new bugs when fixing the allocator code then that's fine, but maybe we should have a github issue to track this and clean up allocator code. Thoughts?

Reply:
my point is that drive-by modifications are dangerous, if anything it should be done in a followup PR

Reply:
> . If @aojea wants to retain the current state to be sure he doesnt introduce new bugs when fixing the allocator code then that's fine, but maybe we should have a github issue to track this and clean up allocator code. Thoughts?

Agree that the allocator needs a better design and definition, but that change require a deeper study of the code to see the implications ... but it is clear out of scope of this PR and dangerous to modify the behaviour here